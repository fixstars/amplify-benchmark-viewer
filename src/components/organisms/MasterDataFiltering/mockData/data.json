[
  { "name": "Tsp" },
  { "name": "Qap" },
  { "name": "MaxCut" },
  { "name": "Sudoku" },
  { "name": "Cvrp" }
]
