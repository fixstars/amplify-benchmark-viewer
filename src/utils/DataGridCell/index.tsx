/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { LinkButton } from 'components/atoms/LinkButton'

const clientColors = [
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
]

const instanceColors = [
  '#ff66ff',
  '#0f66dc',
  '#f57bae',
  '#e75e75',
  '#e5f97c',
  '#1b000d',
  '#f6a695',
  '#692981',
  '#3cde29',
  '#591414',
  '#d66594',
  '#2d884a',
  '#400763',
  '#9d19f4',
  '#fcdeb9',
  '#718de6',
  '#1f4c48',
  '#909743',
  '#2dc1c8',
  '#c17913',
  '#b2a1f6',
  '#9a916d',
  '#406a0f',
  '#004b31',
  '#18729f',
  '#9997c3',
  '#9cf0ec',
  '#ce843c',
  '#2d326e',
  '#76ed60',
  '#d7f7e8',
  '#bb15d7',
  '#87a7d8',
  '#94053a',
  '#e5c8a6',
  '#e2f2c0',
]

export interface ProblemInstances {
  readonly [key: string]: ReadonlyArray<string> | undefined
}

const renderClientsCell = (key: string, clients: ReadonlyArray<string>) => {
  const getBackgroundColor = (client: string): string => {
    const cache = JSON.parse(localStorage.getItem('clientColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('clientColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[client]

    if (backgroundColor === undefined) {
      const length = clientColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = clientColors[cacheIndex]
      cache[client] = backgroundColor
      localStorage.setItem('clientColorCache', JSON.stringify(cache))
      localStorage.setItem('clientColorIndexCache', JSON.stringify(cacheIndex))
      return backgroundColor
    }

    return backgroundColor
  }

  return clients.map((client) => (
    <div key={`${key}_${client}`} style={{ marginRight: 10 }}>
      <LinkButton
        label={client}
        link={`/clients/${client}`}
        backgroundColor={getBackgroundColor(client)}
      />
    </div>
  ))
}

const renderProblemInstancesCell = (
  key: string,
  problemInstances: ProblemInstances,
) => {
  const getBackgroundColor = (instance: string): string => {
    const cache = JSON.parse(localStorage.getItem('instanceColorCache') ?? '{}')
    let cacheIndex = JSON.parse(
      localStorage.getItem('instanceColorIndexCache') ?? '-1',
    )
    let backgroundColor = cache[instance]

    if (backgroundColor === undefined) {
      const length = instanceColors.length
      cacheIndex = cacheIndex >= length ? 0 : cacheIndex + 1
      backgroundColor = instanceColors[cacheIndex]
      cache[instance] = backgroundColor
      localStorage.setItem('instanceColorCache', JSON.stringify(cache))
      localStorage.setItem(
        'instanceColorIndexCache',
        JSON.stringify(cacheIndex),
      )
      return backgroundColor
    }

    return backgroundColor
  }
  const pairs = Object.entries(problemInstances)
  pairs.sort(function (p1, p2) {
    const p1Key = p1[0]
    const p2Key = p2[0]
    if (p1Key < p2Key) {
      return -1
    }
    if (p1Key > p2Key) {
      return 1
    }
    return 0
  })
  problemInstances = Object.fromEntries(pairs)
  return Object.keys(problemInstances).map((className) => (
    <div
      key={`${key}_${className}`}
      style={{ margin: '5px 0 5px 0', display: 'flex' }}
    >
      <span style={{ marginRight: 10 }}>{className}:</span>
      <div style={{ display: 'flex', flexWrap: 'wrap' }}>
        {problemInstances[className]?.map((instance) => (
          <div
            key={`${key}_${className}_${instance}`}
            style={{ marginRight: 10, marginBottom: 10 }}
          >
            <LinkButton
              label={instance}
              link={`/classes/${className}/instances/${instance}`}
              backgroundColor={getBackgroundColor(instance)}
            />
          </div>
        ))}
      </div>
    </div>
  ))
}

export { renderClientsCell, renderProblemInstancesCell }
