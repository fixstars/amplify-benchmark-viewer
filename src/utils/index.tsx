export * from './DataGridCell'
export * from './statistics'
