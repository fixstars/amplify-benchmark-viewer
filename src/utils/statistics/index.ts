function quartile(data: readonly number[], q: number) {
  const sortedData = data.slice().sort((a, b) => a - b)
  const pos = Math.floor((sortedData.length - 1) * q)
  return sortedData[pos]
}
function quartile0(data: readonly number[]) {
  return quartile(data, 0)
}
function quartile25(data: readonly number[]) {
  return quartile(data, 0.25)
}

function quartile50(data: readonly number[]) {
  return quartile(data, 0.5)
}

function quartile75(data: readonly number[]) {
  return quartile(data, 0.75)
}

function quartile100(data: readonly number[]) {
  return quartile(data, 1)
}

export { quartile0, quartile25, quartile50, quartile75, quartile100 }
