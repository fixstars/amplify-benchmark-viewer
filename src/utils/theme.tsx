/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { createTheme } from '@mui/material/styles'

export const theme = createTheme({
  palette: {
    primary: {
      main: '#111',
    },
  },
  typography: {
    h2: {
      fontSize: '1.5rem',
    },
  },
  components: {
    MuiTooltip: {
      styleOverrides: {
        tooltip: {
          backgroundColor: '#111',
          color: '#f4f1f4',
          font: '10px',
        },
      },
    },
  },
})
