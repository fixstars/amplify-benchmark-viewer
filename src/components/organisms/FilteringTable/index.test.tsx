/**
 * Copyright (c) Fixstars Corporation and Fixstars Amplify Corporation.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
import { render, fireEvent, screen } from '@testing-library/react'
import { mockDataGrid } from 'utils/test'

import { FilteringTable } from '.'
import mockData from './mockData/data.json'

describe('<FilteringTable />', () => {
  it('Rendered well', async () => {
    const { container } = render(
      <FilteringTable
        title="Problem Class"
        data={mockData}
        selectedData={[]}
      />,
    )

    const dataGrid = mockDataGrid.mock.calls[0][0]
    expect(dataGrid.sx).toEqual({
      '&.MuiDataGrid-root .MuiDataGrid-cell:focus-within': {
        cursor: 'pointer',
        outline: 'none !important',
      },
      '&.MuiDataGrid-root .MuiDataGrid-cellContent': {
        cursor: 'pointer',
      },
    })
    expect(dataGrid.rows).toEqual([
      { name: 'Cvrp' },
      { name: 'MaxCut' },
      { name: 'Qap' },
      { name: 'Sudoku' },
      { name: 'Tsp' },
    ])
    expect(dataGrid.columns).toEqual([
      {
        field: 'name',
        headerName: 'Problem Class',
        flex: 1,
        sortable: false,
      },
    ])
    expect(dataGrid.initialState).toEqual({
      pagination: {
        pageSize: mockData.length,
      },
    })
    expect(dataGrid.getRowHeight.name).toBe('getRowHeight')
    expect(dataGrid.getRowId.name).toBe('getRowId')
    expect(dataGrid.hideFooter).toBe(true)
    expect(dataGrid.disableColumnMenu).toBe(true)
    expect(dataGrid.checkboxSelection).toBe(true)
    expect(dataGrid.onSelectionModelChange.name).toBe('onSelectionModelChange')

    expect(container).toMatchSnapshot()
  })

  it('onSelected', async () => {
    const handleSelect = jest.fn()

    render(
      <FilteringTable
        title="Problem Class"
        data={mockData}
        selectedData={[]}
        onSelected={handleSelect}
      />,
    )

    expect(handleSelect.mock.calls.length).toBe(0)

    fireEvent.click(screen.getByText('Cvrp'))
    expect(handleSelect.mock.calls.length).toBe(1)
    expect(handleSelect.mock.calls[0][0]).toEqual(['Cvrp'])
  })

  it('selected data', async () => {
    render(
      <FilteringTable
        title="Problem Class"
        data={mockData}
        selectedData={['Cvrp', 'MaxCut']}
      />,
    )

    const dataGrid = mockDataGrid.mock.calls[0][0]
    expect(dataGrid.selectionModel).toEqual(['Cvrp', 'MaxCut'])
  })
})
