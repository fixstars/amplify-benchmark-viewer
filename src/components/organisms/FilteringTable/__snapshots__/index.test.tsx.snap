// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilteringTable /> Rendered well 1`] = `
<div>
  <div
    aria-colcount="2"
    aria-multiselectable="false"
    aria-rowcount="6"
    class="MuiDataGrid-root MuiDataGrid-root--densityStandard MuiDataGrid-withBorderColor css-1ftu7fc-MuiDataGrid-root"
    role="grid"
  >
    <div
      class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
    >
      <div
        class="MuiDataGrid-columnHeaders MuiDataGrid-withBorderColor css-1iyq7zh-MuiDataGrid-columnHeaders"
        role="presentation"
        style="min-height: 56px; max-height: 56px; line-height: 56px;"
      >
        <div
          class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-gl260s-MuiDataGrid-columnHeadersInner"
          role="rowgroup"
          style="transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-rowindex="1"
            class="css-yrdy0g-MuiDataGrid-columnHeaderRow"
            role="row"
          >
            <div
              aria-colindex="1"
              aria-sort="none"
              class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-withBorderColor MuiDataGrid-columnHeaderCheckbox"
              data-field="__check__"
              role="columnheader"
              style="height: 56px; width: 50px; min-width: 50px; max-width: 50px;"
              tabindex="0"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
                role="presentation"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                  role="presentation"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <span
                      class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        aria-label="Select all rows"
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        tabindex="-1"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
            <div
              aria-colindex="2"
              aria-label="Problem Class"
              aria-sort="none"
              class="MuiDataGrid-columnHeader MuiDataGrid-withBorderColor"
              data-field="name"
              role="columnheader"
              style="height: 56px; width: 0px; min-width: 0; max-width: 0;"
              tabindex="-1"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
                role="presentation"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                  role="presentation"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-t89xny-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      Problem Class
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiDataGrid-virtualScroller css-axafay-MuiDataGrid-virtualScroller"
        role="presentation"
        style="overflow-x: hidden;"
      >
        <div
          class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-1kwdphh-MuiDataGrid-virtualScrollerContent"
          role="presentation"
          style="width: auto; height: 260px; min-height: auto;"
        >
          <div
            class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
            role="rowgroup"
            style="transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-rowindex="2"
              aria-selected="false"
              class="MuiDataGrid-row MuiDataGrid-row--dynamicHeight"
              data-id="Cvrp"
              data-rowindex="0"
              role="row"
              style="max-height: none; min-height: auto;"
            >
              <div
                aria-colindex="1"
                aria-colspan="1"
                class="MuiDataGrid-cellCheckbox MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textCenter MuiDataGrid-withBorderColor"
                data-colindex="0"
                data-field="__check__"
                role="cell"
                style="min-width: 50px; max-width: 50px; min-height: auto; max-height: none;"
                tabindex="-1"
              >
                <span
                  class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    aria-label="Select row"
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    tabindex="-1"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
              </div>
              <div
                aria-colindex="2"
                aria-colspan="1"
                class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                data-colindex="1"
                data-field="name"
                role="cell"
                style="min-width: 0; max-width: 0; min-height: auto; max-height: none;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-cellContent"
                  role="presentation"
                  title="Cvrp"
                >
                  Cvrp
                </div>
              </div>
            </div>
            <div
              aria-rowindex="3"
              aria-selected="false"
              class="MuiDataGrid-row MuiDataGrid-row--dynamicHeight"
              data-id="MaxCut"
              data-rowindex="1"
              role="row"
              style="max-height: none; min-height: auto;"
            >
              <div
                aria-colindex="1"
                aria-colspan="1"
                class="MuiDataGrid-cellCheckbox MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textCenter MuiDataGrid-withBorderColor"
                data-colindex="0"
                data-field="__check__"
                role="cell"
                style="min-width: 50px; max-width: 50px; min-height: auto; max-height: none;"
                tabindex="-1"
              >
                <span
                  class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    aria-label="Select row"
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    tabindex="-1"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
              </div>
              <div
                aria-colindex="2"
                aria-colspan="1"
                class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                data-colindex="1"
                data-field="name"
                role="cell"
                style="min-width: 0; max-width: 0; min-height: auto; max-height: none;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-cellContent"
                  role="presentation"
                  title="MaxCut"
                >
                  MaxCut
                </div>
              </div>
            </div>
            <div
              aria-rowindex="4"
              aria-selected="false"
              class="MuiDataGrid-row MuiDataGrid-row--dynamicHeight"
              data-id="Qap"
              data-rowindex="2"
              role="row"
              style="max-height: none; min-height: auto;"
            >
              <div
                aria-colindex="1"
                aria-colspan="1"
                class="MuiDataGrid-cellCheckbox MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textCenter MuiDataGrid-withBorderColor"
                data-colindex="0"
                data-field="__check__"
                role="cell"
                style="min-width: 50px; max-width: 50px; min-height: auto; max-height: none;"
                tabindex="-1"
              >
                <span
                  class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiCheckbox-root MuiCheckbox-colorPrimary MuiCheckbox-sizeMedium MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                >
                  <input
                    aria-label="Select row"
                    class="PrivateSwitchBase-input css-1m9pwf3"
                    data-indeterminate="false"
                    tabindex="-1"
                    type="checkbox"
                  />
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="CheckBoxOutlineBlankIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
              </div>
              <div
                aria-colindex="2"
                aria-colspan="1"
                class="MuiDataGrid-cell MuiDataGrid-cell--textLeft MuiDataGrid-withBorderColor"
                data-colindex="1"
                data-field="name"
                role="cell"
                style="min-width: 0; max-width: 0; min-height: auto; max-height: none;"
                tabindex="-1"
              >
                <div
                  class="MuiDataGrid-cellContent"
                  role="presentation"
                  title="Qap"
                >
                  Qap
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
